### PR TITLE
feat: add opentelemetry usage example

### PR DESCRIPTION
**Pull request by daviderli614**
## What's changed and what's your intention?
When after https://github.com/GreptimeTeam/greptimedb-ingester-go/pull/39 merge, we can provide some example to user.

## Checklist

- [ ]  I have added the necessary unit tests and integration tests.

## Refer to a related PR or issue link (optional)

### CODE REVIEW

**Comment by sunng87** [comment]
by the way, is there any concept like feature gate or optional dependency in go? I would suggest to make otel deps optional so we keep the core sdk compact.

**Comment by daviderli614** [comment]
> by the way, is there any concept like feature gate or optional dependency in go? I would suggest to make otel deps optional so we keep the core sdk compact.

Yes, a [switch](https://github.com/GreptimeTeam/greptimedb-ingester-go/pull/47/files#diff-f3d8c65178a58b3b49e68ea350dc50fe7ffca821d97690473117ee0b0a5e3b6cR108-R110) option has been added to the configuration. The default value is false, which means it is closed. If the user needs to turn it on, it can be set to true.
```
cfg := greptime.NewConfig(host).WithDatabase(database).
  WithTraceProvider(tracerProvider).WithTracesEnabled(true).
  WithMeterProvider(meterProvider).WithMetricsEnabled(true)
```

Also, the deps [go.mod](https://github.com/GreptimeTeam/greptimedb-ingester-go/blob/1399b2df82adacc3009532fe6d04060257839eb5/examples/opentelemetry/go.mod) is set here so as not to affect the project's main [go.mod](https://github.com/GreptimeTeam/greptimedb-ingester-go/blob/main/go.mod) and not to pollute the contents of the main go.mod.

File: examples/opentelemetry/main.go [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+func main() {
+	log.Printf("Waiting for connection...")
+
+	ctx, cancel := signal.NotifyContext(context.Background(), os.Interrupt)
+	defer cancel()
+
+	conn, err := initConn()
+	if err != nil {
+		log.Fatal(err)
+		return
```
Comment (zyy17) [inline examples/opentelemetry/main.go:70]: `return` is unnecessary because of the use of  `log.Fatal()`.

File: examples/opentelemetry/main.go [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+
+	greptime "github.com/GreptimeTeam/greptimedb-ingester-go"
+	"github.com/GreptimeTeam/greptimedb-ingester-go/table"
+	"github.com/GreptimeTeam/greptimedb-ingester-go/table/types"
+)
+
+var (
+	client      *greptime.Client
```
Comment (zyy17) [inline examples/opentelemetry/main.go:47]: suggestion: You don't have to take `client` as a global variable. You can pass it as an argument in a function.

File: examples/opentelemetry/main.go [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+	serviceName = semconv.ServiceNameKey.String("test-otel")
```
Comment (zyy17) [inline examples/opentelemetry/main.go:48]: `serviceName` also does not need to be a global variable.

File: examples/opentelemetry/main.go [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+	)
+	otel.SetMeterProvider(meterProvider)
+
+	return nil
+}
+
+func serveMetrics() {
+	log.Printf("serving metrics at localhost:2233/metrics")
+	http.Handle("/metrics", promhttp.Handler())
+	err := http.ListenAndServe(":2233", nil) //nolint:gosec // Ignoring G114: Use of net/http serve function that has no support for setting timeouts.
```
Comment (zyy17) [inline examples/opentelemetry/main.go:185]: The `:2233` can be extracted as a const variable.

File: examples/opentelemetry/main.go [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+	"github.com/GreptimeTeam/greptimedb-ingester-go/table/types"
+)
+
+var (
+	client      *greptime.Client
+	serviceName = semconv.ServiceNameKey.String("test-otel")
+
+	// It connects the OpenTelemetry Collector through local gRPC connection.
+	// You may replace `localhost:4317` with your endpoint.
+	tracingEndpoint = "localhost:4317"
```
Comment (zyy17) [inline examples/opentelemetry/main.go:52]: Those string can be const variables.

File: examples/opentelemetry/main.go [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+	}
+
+	if err := dtbl.AddRow(3, "hello", 3.3, time3); err != nil {
+		log.Println(err)
+	}
+
+	return []*table.Table{itbl, utbl, dtbl}
+}
+
+func write(data *table.Table) {
```
Comment (zyy17) [inline examples/opentelemetry/main.go:275]: you can pass `client` to `write()`.

**Comment by daviderli614** [comment]
Modify are complete and ready for review @zyy17

File: examples/opentelemetry/main.go [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+	log.Printf("serving metrics at localhost%s/metrics", metricsBindAddr)
+	http.Handle("/metrics", promhttp.Handler())
+	err := http.ListenAndServe(metricsBindAddr, nil) //nolint:gosec // Ignoring G114: Use of net/http serve function that has no support for setting timeouts.
+	if err != nil {
+		fmt.Printf("error serving http: %v", err)
+		return
+	}
+}
+
+func initData() []*table.Table {
```
Comment (zyy17) [inline examples/opentelemetry/main.go:201]: It's not a best practice to log all error messages. We can propagate the error to the caller.

```suggestion
func initData() ([]*table.Table, error) {
}
```

File: examples/opentelemetry/main.go [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+
+	return nil
+}
+
+func serveMetrics() {
+	log.Printf("serving metrics at localhost%s/metrics", metricsBindAddr)
+	http.Handle("/metrics", promhttp.Handler())
+	err := http.ListenAndServe(metricsBindAddr, nil) //nolint:gosec // Ignoring G114: Use of net/http serve function that has no support for setting timeouts.
+	if err != nil {
+		fmt.Printf("error serving http: %v", err)
```
Comment (zyy17) [inline examples/opentelemetry/main.go:196]: Why not use `log` package?

File: examples/opentelemetry/main.go [hunk 9]
```diff
@@ -0,0 +1,292 @@
...
+
+	exporter, err := otelprom.New(otelprom.WithNamespace("greptime"))
+	if err != nil {
+		log.Fatal(err)
+	}
+	meterProvider := metric.NewMeterProvider(
+		metric.WithReader(exporter),
+	)
+
+	err = initMeterProvider(ctx, res, conn)
```
Comment (zyy17) [inline examples/opentelemetry/main.go:118]:
```suggestion
	if err = initMeterProvider(ctx, res, conn); err != nil { ... }
```

File: examples/opentelemetry/main.go [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+}
+
+func newClient(tracerProvider trace.TracerProvider, meterProvider *metric.MeterProvider) *client {
+	cfg := greptime.NewConfig(host).WithDatabase(database).
+		WithTraceProvider(tracerProvider).WithTracesEnabled(true).
+		WithMeterProvider(meterProvider).WithMetricsEnabled(true)
+
+	gtClient, err := greptime.NewClient(cfg)
+	if err != nil {
+		log.Fatal(err)
```
Comment (zyy17) [inline examples/opentelemetry/main.go:73]: You can propagate the error here. For example:

```
func newClient(tracerProvider trace.TracerProvider, meterProvider *metric.MeterProvider) (*client, error) { ... }
```

File: examples/opentelemetry/main.go [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+	"go.opentelemetry.io/otel/trace"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+
+	greptime "github.com/GreptimeTeam/greptimedb-ingester-go"
+	"github.com/GreptimeTeam/greptimedb-ingester-go/table"
+	"github.com/GreptimeTeam/greptimedb-ingester-go/table/types"
+)
+
+var (
```
Comment (zyy17) [inline examples/opentelemetry/main.go:46]: Why not use the single const block to manage the default settings? For example:

```
const (
  ...
)
```

File: examples/opentelemetry/main.go [hunk 12]
```diff
@@ -0,0 +1,292 @@
...
+	}
+
+	if err := dtbl.AddRow(3, "hello", 3.3, time3); err != nil {
+		log.Println(err)
+	}
+
+	return []*table.Table{itbl, utbl, dtbl}
+}
+
+func (c client) write(data *table.Table) {
```
Comment (zyy17) [inline examples/opentelemetry/main.go:284]: Ditto. You can propagate the error to the caller.

**Review by zyy17 — APPROVED** [review]

**Pull request merged**